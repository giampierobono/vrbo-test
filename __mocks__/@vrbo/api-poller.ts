import { mockState } from "../../jest-global-mocks";

const apiPoller = jest.requireActual("@vrbo/api-poller");

apiPoller.poll = jest.fn(() =>
  Promise.resolve(mockState.propertiesList.properties)
);

module.exports = apiPoller;
