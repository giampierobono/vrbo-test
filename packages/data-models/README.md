# @vrbo/data-models

This packages contains all interfaces, enums and configs shared with all other packages. 

Below details of what's included in this package.

## Configs

| Name                     | Type                              | Description                                                  |
| :----------------------- | --------------------------------- | ------------------------------------------------------------ |
| `propertiesSortByConfig` | `PropertiesSortByConfigModelType` | Configuration for sortBy selectbox. Introduced to simplify adding new way of sort properties list without changing any other code. |

## Custom types

| Name                              | Decription                                                   |
| --------------------------------- | ------------------------------------------------------------ |
| `PropertiesSortByConfigModelType` | Defines a mapped object having as key one of `PropertiesSortByLabels` keys and a `PropertiesSortByConfig` as value. |

## Enums

| Name                     | Decription                                                   |
| ------------------------ | ------------------------------------------------------------ |
| `PropertiesSortBy`       | Defines a set of values representing `Property` prop value names to be used to sort properties list. |
| `PropertiesSortByLabels` | Define a set of static labels to be used as sortBy selectbox options. Imagine having a translation system, this enum could contain translation key values. |
| `PropertyType`           | Define type of properties.                                   |

## Interfaces

| Name                     | Decription                                                   |
| ------------------------ | ------------------------------------------------------------ |
| `ActionWithPayload<T>`   | Defines an object exteding `rxjs/Action` with a typed (`T`) payload property. It has been included to have typed redux actions payload. Ex: `const myAction: ActionWithPayload<Property[]>` |
| `PollerConfig`           | Define a config object to be sent as `poll` method input.    |
| `Properties`             | Define an object containing an array of `Property`. Not used. |
| `PropertiesSortByConfig` | Define an object used to configure the sortBy selectbox.     |
| `Property`               | Define a object describing a property (both house and condos). |

