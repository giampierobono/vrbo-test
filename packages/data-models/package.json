{
  "name": "@vrbo/data-models",
  "version": "0.0.1",
  "license": "MIT",
  "author": "Giampiero Bono <giampiero.bono@hotmail.it>",
  "files": [
    "lib"
  ],
  "main": "./lib/data-models.js",
  "types": "./lib/types.d.ts",
  "scripts": {
    "prebuild": "npm run clean:lib",
    "build": "tsc",
    "clean:lib": "rimraf lib/",
    "clean:node_modules": "rimraf node_modules",
    "lint": "tslint -c tslint.json 'src/**/*.ts'",
    "lint:fix": "npm run lint -- --fix",
    "sort-package-json": "sort-package-json package.json"
  },
  "devDependencies": {
    "ts-loader": "^6.2.1",
    "tslint": "^5.20.1",
    "typescript": "^3.4.1"
  },
  "peerDependencies": {
    "rimraf": "^3.0.0"
  }
}
