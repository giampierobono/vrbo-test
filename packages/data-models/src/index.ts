export * from "./configs";
export * from "./enums";
export * from "./models";
export * from "./custom-types";
