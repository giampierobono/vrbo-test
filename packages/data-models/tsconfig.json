{
  "compilerOptions": {
    "baseUrl": ".",
    "module": "commonjs",
    "target": "es5",
    "lib": [ "es2015", "dom" ],
    "outDir": "lib",
    "sourceMap": true,
    "skipLibCheck": true,
    "moduleResolution": "node",
    "rootDir": "src",
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "importHelpers": true,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true,
    "noUnusedLocals": false,
    "declaration": true,
    "declarationMap": true,
    "allowSyntheticDefaultImports": true
  },
  "exclude": [
    "lib",
    "node_modules"
  ]
}
