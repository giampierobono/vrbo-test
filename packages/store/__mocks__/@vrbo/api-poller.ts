// tslint:disable-next-line:no-var-requires
const apiPollerStore = require("../../../../__mocks__/@vrbo/api-poller");

module.exports = apiPollerStore;
