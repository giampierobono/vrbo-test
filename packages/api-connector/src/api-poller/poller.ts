import { PollerConfig, Property, PropertyType } from "@vrbo/data-models";
import { getRandomNumber } from "../utils";

const pollerDefaultConfig: PollerConfig = {
  limit: 10,
  type: PropertyType.Houses
};

const fakeDB: { houses: string[]; condos: string[] } = {
  houses: [
    "Luxury Home Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod",
    "Luxury Home tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam",
    "Luxury Home quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo",
    "Luxury Home consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse",
    "Luxury Home cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non",
    "Luxury Home exercitation ullamco laboris nisi ut aliquip ullamco laboris",
    "Luxury Home deserunt mollit anim id est laborum",
    "Luxury Home mollit anim id est laborum laboris nisi ut aliquip ex ea commodo",
    "Luxury Home qui officia deserunt mollit anim id est laborum amet, consectetur adipisicing elit, sed do eiusmod",
    "Luxury Home ipsum dolor sit amet, consectetur deserunt mollit anim id est laborum"
  ],
  condos: [
    "Fancy Condo Bacon ipsum dolor amet pastrami chicken venison, meatball alcatra pork belly short ribs",
    "Fancy Condo Short loin alcatra tail beef burgdoggen hamburger spare ribs meatball kevin sirloin leberkas pork loin prosciutto shankle bresaola",
    "Fancy Condo Sirloin jerky landjaeger leberkas jowl tongue salami capicola sausage cow hamburger picanha ball tip meatball corned beef",
    "Fancy Condo Leberkas fatback meatloaf kielbasa bresaola biltong jowl turkey sausage rump",
    "Fancy Condo Pork chop chuck shank jowl landjaeger",
    "Fancy Condo Ground loin alcatra tail beef burgdoggen hamburger cupim turducken ham hock venison kielbasa",
    "Fancy Condo cupim turducken ham hock venison kielbasa tail beef burgdoggen hamburger",
    "Fancy Condo venison kielbasa beef burgdoggen hamburger spare",
    "Fancy Condo jerky landjaeger leberkas, cupim turducken ham hock venison kielbasa",
    "Fancy Condo sirloin capicola, Short loin alcatra tail beef venison kielbasa"
  ]
};

const getData = (type: PropertyType): Property[] => {
  let item;
  let i;
  let len;
  const list = fakeDB[type] || [];
  const results = [];

  for (i = 0, len = list.length; i < len; i++) {
    item = list[i];

    results.push({
      title: item,
      rating: getRandomNumber(0, 5),
      ratingCount: Math.floor(Math.random() * 50) + 1,
      price: Math.floor(Math.random() * 300) + 100,
      beds: Math.floor(Math.random() * 5) + 1,
      bath: Math.floor(Math.random() * 5) + 1,
      sqFeet: Math.floor(Math.random() * 3000) + 1500
    });
  }
  return results;
};

const processData = (data: Property[], limit: number): Property[] =>
  data.slice(0, limit);

export const poll = (
  options: PollerConfig,
  callBack?: (properties: Property[]) => void
) => {
  const config: PollerConfig = { ...pollerDefaultConfig, ...options };
  return new Promise(resolve => {
    setTimeout(() => {
      const payload = processData(getData(config.type), config.limit);

      if (callBack) {
        callBack(payload);
      }
      resolve(payload);
    }, getRandomNumber(400, 2000));
  });
};
