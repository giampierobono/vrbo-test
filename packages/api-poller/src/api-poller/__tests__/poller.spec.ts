import { Property, PropertyType } from "@vrbo/data-models";
import { poll } from "../poller";

describe("api poller", () => {
  let mathRandomSpy: jest.SpyInstance;

  beforeEach(() => {
    mathRandomSpy = jest.spyOn(Math, "random").mockReturnValue(1);
  });

  afterEach(() => {
    mathRandomSpy.mockReset();
  });

  it("should return list of 10 houses", async done => {
    const result = poll({ limit: 10, type: PropertyType.Houses });
    await result.then((houses: Property[]) => {
      expect(houses.length).toBe(10);
      expect(houses).toMatchSnapshot();
      done();
    });
  });

  it("should return list of 5 houses", async done => {
    const result = poll({ limit: 5, type: PropertyType.Houses });
    await result.then((houses: Property[]) => {
      expect(houses.length).toBe(5);
      expect(houses).toMatchSnapshot();
      done();
    });
  });

  it("should return list of 10 condos", async done => {
    const result = poll({ limit: 10, type: PropertyType.Condos });
    await result.then((condos: Property[]) => {
      expect(condos.length).toBe(10);
      expect(condos).toMatchSnapshot();
      done();
    });
  });

  it("should return list of 5 condos", async done => {
    const result = poll({ limit: 5, type: PropertyType.Condos });
    await result.then((condos: Property[]) => {
      expect(condos.length).toBe(5);
      expect(condos).toMatchSnapshot();
      done();
    });
  });
});
