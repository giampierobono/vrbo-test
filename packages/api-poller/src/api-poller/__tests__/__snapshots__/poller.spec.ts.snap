// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api poller should return list of 5 condos 1`] = `
Array [
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Bacon ipsum dolor amet pastrami chicken venison, meatball alcatra pork belly short ribs",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Short loin alcatra tail beef burgdoggen hamburger spare ribs meatball kevin sirloin leberkas pork loin prosciutto shankle bresaola",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Sirloin jerky landjaeger leberkas jowl tongue salami capicola sausage cow hamburger picanha ball tip meatball corned beef",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Leberkas fatback meatloaf kielbasa bresaola biltong jowl turkey sausage rump",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Pork chop chuck shank jowl landjaeger",
  },
]
`;

exports[`api poller should return list of 5 houses 1`] = `
Array [
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non",
  },
]
`;

exports[`api poller should return list of 10 condos 1`] = `
Array [
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Bacon ipsum dolor amet pastrami chicken venison, meatball alcatra pork belly short ribs",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Short loin alcatra tail beef burgdoggen hamburger spare ribs meatball kevin sirloin leberkas pork loin prosciutto shankle bresaola",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Sirloin jerky landjaeger leberkas jowl tongue salami capicola sausage cow hamburger picanha ball tip meatball corned beef",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Leberkas fatback meatloaf kielbasa bresaola biltong jowl turkey sausage rump",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Pork chop chuck shank jowl landjaeger",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo Ground loin alcatra tail beef burgdoggen hamburger cupim turducken ham hock venison kielbasa",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo cupim turducken ham hock venison kielbasa tail beef burgdoggen hamburger",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo venison kielbasa beef burgdoggen hamburger spare",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo jerky landjaeger leberkas, cupim turducken ham hock venison kielbasa",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Fancy Condo sirloin capicola, Short loin alcatra tail beef venison kielbasa",
  },
]
`;

exports[`api poller should return list of 10 houses 1`] = `
Array [
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home exercitation ullamco laboris nisi ut aliquip ullamco laboris",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home deserunt mollit anim id est laborum",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home mollit anim id est laborum laboris nisi ut aliquip ex ea commodo",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home qui officia deserunt mollit anim id est laborum amet, consectetur adipisicing elit, sed do eiusmod",
  },
  Object {
    "bath": 6,
    "beds": 6,
    "price": 400,
    "rating": 5,
    "ratingCount": 51,
    "sqFeet": 4500,
    "title": "Luxury Home ipsum dolor sit amet, consectetur deserunt mollit anim id est laborum",
  },
]
`;
